sql/sql_table.cc

```diff
@@ -10130,6 +10130,21 @@ bool mysql_alter_table(THD *thd, const LEX_CSTRING *new_db,
     DBUG_RETURN(true);
   }
 
+  /*
+    MDEV-23485: Change table to merge engine may cause table data lost.
+
+    If we want to change a table to merge engine, the original table
+    should be merge engine table, otherwise may cause table data lost.
+  */
+  if (ha_check_storage_engine_flag(create_info->db_type, HA_NO_COPY_ON_ALTER) &&
+      table->file->ht->db_type != create_info->db_type->db_type)
```

Comment:
I think it'd be nice not to enforce this restriction for empty tables.

Also, since `HA_NO_COPY_ON_ALTER` is a handler flag and handler is not going to be available until later in the code, you'd have to either use `get_new_handler()` here or move the fix under `else` branch of `if (!(new_table->file->ha_table_flags() & HA_NO_COPY_ON_ALTER))` condition.